engineapi/proxy.go

```diff
@@ -19,77 +23,296 @@ import (
 )
 
 const (
-	ProxyPollInterval = 5 * time.Second
+	MaxProxyClientConcurrentInUse = 1000
+	DefaultProxyClientPerImage    = 3
+	MaxProxyClientPerImage        = 3 * DefaultProxyClientPerImage
+	ProxyPollInterval             = 5 * time.Second
+	ProxyClientMaxRetry           = 10
+	ProxyGCMaxRetry               = 3
 )
 
-type EngineClientProxy interface {
-	EngineClient
-	BackupTargetBinaryClient
-
-	IsGRPC() bool
-	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
-	Stop(string) error
-	Ping() error
-}
-
 type EngineClientProxyHandler struct {
-	Clients map[string]EngineClientProxy
+	handle map[string]EngineClientProxys
 
 	logger logrus.FieldLogger
+	ds     *datastore.DataStore
 
-	ds *datastore.DataStore
+	stopCh chan struct{}
 
 	lock *sync.RWMutex
+
+	clock        clock.Clock
+	resyncPeriod time.Duration
+}
+
+type EngineClientProxys struct {
+	Clients map[int]EngineClientProxy
+
+	instanceManagerName string
 }
 
-func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore) *EngineClientProxyHandler {
+func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore, resyncPeriod time.Duration, stopCh chan struct{}) *EngineClientProxyHandler {
 	return &EngineClientProxyHandler{
-		Clients: make(map[string]EngineClientProxy),
-		logger:  logger,
-		ds:      ds,
-		lock:    &sync.RWMutex{},
+		handle:       make(map[string]EngineClientProxys),
+		logger:       logger,
+		ds:           ds,
+		stopCh:       stopCh,
+		lock:         &sync.RWMutex{},
+		clock:        &clock.RealClock{},
+		resyncPeriod: resyncPeriod,
 	}
 }
 
-func getLoggerForProxy(logger logrus.FieldLogger, image string) *logrus.Entry {
+func getLoggerForProxyHandler(logger logrus.FieldLogger, im *longhorn.InstanceManager) *logrus.Entry {
 	return logger.WithFields(
 		logrus.Fields{
-			"instanceManagerImage": image,
+			"instanceManager": im.Name,
+			"image":           im.Spec.Image,
 		},
 	)
 }
 
+func (h *EngineClientProxyHandler) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	h.logger.Info("Start Longhorn proxy manager")
```

Comment:
"Starting engine client proxy handler"? Suggest having consistent names corresponding to the method, structure, etc.

Reply:
Updated.